### PR TITLE
Add OAuth example for Product Hunt

### PR DESCRIPTION
**Pull request by hhua**
Official API documents is [here](https://api.producthunt.com/v1/docs/oauth_user_authentication/oauth_authorize_ask_for_access_grant_code_on_behalf_of_the_user)

### CODE REVIEW

**Comment by kullfar** [comment]
Hi, good job!
Could you provide Example using this new API (as here https://github.com/scribejava/scribejava/tree/master/scribejava-apis/src/test/java/com/github/scribejava/apis/examples)

File: scribejava-apis/src/main/java/com/github/scribejava/ProductHuntApi.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import com.github.scribejava.core.extractors.AccessTokenExtractor;
+import com.github.scribejava.core.extractors.JsonTokenExtractor;
+import com.github.scribejava.core.model.OAuthConfig;
+import com.github.scribejava.core.model.OAuthConstants;
+import com.github.scribejava.core.model.Verb;
+import com.github.scribejava.core.oauth.OAuthService;
+import com.github.scribejava.core.utils.OAuthEncoder;
+
+public class ProductHuntApi extends DefaultApi20 {
+    private static final String AUTHORIZE_URL = "https://api.producthunt.com/v1/oauth/authorize?client_id=%s&redirect_uri=%s&response_type=code&scope=public+private";
```
Comment (kullfar) [inline scribejava-apis/src/main/java/com/github/scribejava/ProductHuntApi.java]: scope should be parametrized, shouldn't it?

File: scribejava-apis/src/main/java/com/github/scribejava/ProductHuntApi.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Override
+    public String getAuthorizationUrl(final OAuthConfig config) {
+        Preconditions.checkValidUrl(config.getCallback(),
+                "Must provide a valid url as callback. Product Hunt does not support OOB");
+        final StringBuilder sb = new StringBuilder(String.format(AUTHORIZE_URL, config.getApiKey(), OAuthEncoder.encode(
+                config.getCallback())));
+        if (config.hasScope()) {
+            sb.append('&').append(OAuthConstants.SCOPE).append('=').append(OAuthEncoder.encode(config.getScope()));
```
Comment (kullfar) [inline scribejava-apis/src/main/java/com/github/scribejava/ProductHuntApi.java]: I see, that scope is parametrized actually ;-)
"&scope=public+private" looks like redundant part of AUTHORIZE_URL

File: scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+    public ProductHuntServiceImpl(final DefaultApi20 api, final OAuthConfig config) {
+        super(api, config);
+    }
+    
+    @Override
+    protected <T extends AbstractRequest> T createAccessTokenRequest(final Verifier verifier, T request) {
+        super.createAccessTokenRequest(verifier, request);
+        if (!getConfig().hasGrantType()) {
+            request.addParameter(OAuthConstants.GRANT_TYPE, "authorization_code");
```
Comment (kullfar) [inline scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java]: You have "grant_type=authorization_code" parameter in ACCESS_TOKEN_URL.
looks like redundant as well

File: scribejava-apis/src/main/java/com/github/scribejava/ProductHuntApi.java (hunk 2 above)
Comment (hhua) [inline scribejava-apis/src/main/java/com/github/scribejava/ProductHuntApi.java]: Ah, yes. Good call.

File: scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java (hunk 3 above)
Comment (hhua) [inline scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java]: Yes, I will remove it and add the example.

**Comment by hhua** [comment]
@kullfar Thanks for reviewing this!

File: scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java (hunk 3 above)
Comment (kullfar) [inline scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java]: And provide please any test account to run this Example and test ;-)
just credentials here in description or by email (as you want)
Thanks in advance!

**Comment by kullfar** [comment]
welcome! ;-)

File: scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java (hunk 3 above)
Comment (hhua) [inline scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java]: @kullfar Could you please send me the email address which can be used to send you the credential? Thanks

**Comment by hhua** [comment]
@kullfar Example added and other feedback addressed

File: scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java (hunk 3 above)
Comment (kullfar) [inline scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java]: Hi, sorry for the delay. Was on vacation. Use this one
https://github.com/scribejava/scribejava/blob/master/pom.xml#L48

**Comment by kullfar** [comment]
Btw, add one line to the https://github.com/scribejava/scribejava/blob/d84a75236ad3ef6064a2768448496ddf62b954bf/changelog
under [SNAPSHOT] section.
smth like
* add Product Hunt API

**Comment by kullfar** [comment]
BTW, I have the compilation errors
```
[ERROR] /tmp/scribejava/scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java:[14,26] cannot find symbol
  symbol:   class AbstractRequest
  location: class com.github.scribejava.apis.service.ProductHuntServiceImpl
[ERROR] /tmp/scribejava/scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java:[14,76] cannot find symbol
  symbol:   class Verifier
  location: class com.github.scribejava.apis.service.ProductHuntServiceImpl
[ERROR] /tmp/scribejava/scribejava-apis/src/main/java/com/github/scribejava/ProductHuntApi.java:[29,9] cannot find symbol
  symbol:   variable Preconditions
  location: class com.github.scribejava.apis.ProductHuntApi
[ERROR] /tmp/scribejava/scribejava-apis/src/main/java/com/github/scribejava/apis/service/ProductHuntServiceImpl.java:[17,34] cannot find symbol
  symbol:   variable OAuthConstants
  location: class com.github.scribejava.apis.service.ProductHuntServiceImpl
```
etc.

**Comment by hhua** [comment]
@kullfar Updated

File: scribejava-apis/src/test/java/com/github/scribejava/apis/examples/ProductHunt20Example.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
+package com.github.scribejava.apis.examples;
+
+import com.github.scribejava.apis.ProductHuntApi;
+import com.github.scribejava.core.builder.ServiceBuilder;
+import com.github.scribejava.core.model.*;
```
Comment (kullfar) [inline scribejava-apis/src/test/java/com/github/scribejava/apis/examples/ProductHunt20Example.java]: try to not use star imports pls

File: scribejava-apis/src/test/java/com/github/scribejava/apis/examples/ProductHunt20Example.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        // Trade the Request Token and Verfier for the Access Token
+        System.out.println("Trading the Request Token for an Access Token...");
+        Token accessToken = service.getAccessToken(EMPTY_TOKEN, verifier);
+        System.out.println("Got the Access Token!");
+        System.out.println("(if your curious it looks like this: " + accessToken + " )");
+        System.out.println();
+
+        // Now let's go and ask for a protected resource!
+        System.out.println("Now we're going to access a protected resource...");
+        while(true){
```
Comment (kullfar) [inline scribejava-apis/src/test/java/com/github/scribejava/apis/examples/ProductHunt20Example.java]: really? I do not see how this will be finished.

**Comment by kullfar** [comment]
credentials provided by email is invalid. I'm still waiting for a correct one to run the example

**Comment by kullfar** [comment]
Didn't recieve test user credentials. just closing ;-(

**Pull request closed without merging**